storagenode/pieces/store.go

```diff
@@ -46,42 +47,104 @@ type Info struct {
-	// GetExpired gets orders that are expired and were created before some time
+	// GetExpired gets piece IDs stored with storage format V0 that expire or have expired
+	// before the given time
 	GetExpired(ctx context.Context, expiredAt time.Time, limit int64) ([]ExpiredInfo, error)
+	// ForAllV0PieceIDsOwnedBySatellite executes doForEach for each locally stored piece, stored
+	// with storage format V0 in the namespace of the given satellite. If doForEach returns a
+	// non-nil error, ForAllV0PieceIDsOwnedBySatellite will stop iterating and return the error
+	// immediately.
+	ForAllV0PieceIDsOwnedBySatellite(ctx context.Context, blobStore storage.Blobs, satellite storj.NodeID, doForEach func(StoredPieceAccess) error) error
+}
+
+// V0PieceInfoDBForTest is like V0PieceInfoDB, but adds on the Add() method so
+// that test environments with V0 piece data can be set up.
+type V0PieceInfoDBForTest interface {
+	V0PieceInfoDB
+
+	// Add inserts Info to the database. This is only a valid thing to do, now,
+	// during tests, to replicate the environment of a storage node not yet fully
+	// migrated to V1 storage.
+	Add(context.Context, *Info) error
+}
+
+// StoredPieceAccess allows inspection and manipulation of a piece during iteration with
+// ForAllPieceIDsOwnedBySatellite-type methods
+type StoredPieceAccess interface {
+	storage.StoredBlobAccess
+
+	// PieceID gives the pieceID of the piece
+	PieceID() storj.PieceID
+	// Satellite gives the nodeID of the satellite which owns the piece
+	Satellite() (storj.NodeID, error)
+	// ContentSize gives the size of the piece content (not including the piece header, if
+	// applicable)
+	ContentSize(ctx context.Context) (int64, error)
+	// CreationTime returns the piece creation time as given in the original PieceHash (which is
+	// likely not the same as the file mtime). For non-FormatV0 pieces, this requires opening
+	// the file and unmarshaling the piece header. If exact precision is not required, ModTime()
+	// may be a better solution.
+	CreationTime(ctx context.Context) (time.Time, error)
+	// ModTime returns a less-precise piece creation time than CreationTime, but is generally
+	// much faster. For non-FormatV0 pieces, this gets the piece creation time from to the
+	// filesystem instead of the piece header.
+	ModTime(ctx context.Context) (time.Time, error)
 }
 
 // Store implements storing pieces onto a blob storage implementation.
 type Store struct {
-	log   *zap.Logger
-	blobs storage.Blobs
+	log            *zap.Logger
+	blobs          storage.Blobs
+	v0PieceInfo    V0PieceInfoDB
+	expirationInfo PieceExpirationDB
+
+	// The value of reservedSpace is always added to the return value from the
+	// SpaceUsedForPieces() method.
+	// The reservedSpace field is part of an unfortunate hack that enables testing of low-space
+	// or no-space conditions. It is not (or should not be) used under regular operating
+	// conditions.
+	reservedSpace int64
```

Comment:
I think you can remove all of these test-specific things from this file and put them all in test files. If `StoreForTest` is in a test file instead, you can have it overwrite the `SpaceUsedForPieces` method, which is simply a pass-through to the underlying store, but then adds the `reservedSpace` to it. Then you can run your tests on `StoreForTest` and it should work, I believe.

Reply:
Believe me, I would very much like for that to be the case. Tests far away from this package need to be able to simulate having the store be out of space, and that store needs to be the "real" store object owned by a `storagenode.Peer`.

We could replace all current objects of type `*pieces.Store` with an interface, and have a way to reassign the store object owned by a `storagenode.Peer`, but that would mean a _lot_ more changes than this PR already makes, and as much as I don't like test-specific code in non-test files, I don't think that it would be very much better to make a large interface type with a dozen methods and requiring all non-test code to use it, all for the purpose of enabling some test code.

Reply:
:+1: